components/eam/src/physics/cam/shoc.F90

```diff
@@ -3390,54 +3390,49 @@ subroutine vd_shoc_decomp( &
 
 ! OUTPUT VARIABLES
   ! superdiagonal
-  real(rtype), intent(out) :: ca(shcol,nlev)
-  ! subdiagonal
-  real(rtype), intent(out) :: cc(shcol,nlev)
-  ! 1./(1.+ca(k)+cc(k)-cc(k)*ze(k-1))
-  real(rtype), intent(out) :: denom(shcol,nlev)
-  ! Term in tri-diag. matrix system
-  real(rtype), intent(out) :: ze(shcol,nlev)
+  real(rtype), intent(out) :: du(shcol,nlev)
+  ! Factorized version of subdiagonal
+  real(rtype), intent(out) :: dl(shcol,nlev)
+  ! Factorized version of diagonal
+  real(rtype), intent(out) :: d(shcol,nlev)
 
 ! LOCAL VARIABLES
   integer :: i, k
 
-  ! Determine superdiagonal (ca(k)) and subdiagonal (cc(k)) coeffs of the
-  ! tridiagonal diffusion matrix. The diagonal elements  (cb=1+ca+cc) are
-  ! a combination of ca and cc; they are not required by the solver.
-
+  ! Determine superdiagonal (du) and subdiagonal (dl) coeffs of the
+  ! tridiagonal diffusion matrix.
   do k=nlev-1,1,-1
     do i=1,shcol
-      ca(i,k) = kv_term(i,k+1) * tmpi(i,k+1) * rdp_zt(i,k)
-      cc(i,k+1) = kv_term(i,k+1) * tmpi(i,k+1) * rdp_zt(i,k+1)
+      du(i,k)   = -1._rtype * kv_term(i,k+1) * tmpi(i,k+1) * rdp_zt(i,k)
+      dl(i,k+1) = -1._rtype * kv_term(i,k+1) * tmpi(i,k+1) * rdp_zt(i,k+1)
     enddo
   enddo
 
-  ! The bottom element of the upper diagonal (ca) is zero (not used).
-  ! The subdiagonal (cc) is not needed in the solver.
-
-  ca(:,nlev) = 0._rtype
-
-  ! Calculate e(k). This term is required in the solution of the
-  ! tridiagonal matrix as defined by the implicit diffusion equation.
+  ! The bottom element of the superdiagonal (du) and the top element of
+  ! the subdiagonal (dl) is set to zero (not included in linear system).
+  du(:,nlev) = 0._rtype
+  dl(:,1)    = 0._rtype
 
+  ! The diagonal elements are a combination of du and dl (d=1-du-dl). Surface
+  ! fluxes are applied explicitly in the diagonal at the top level.
   do i=1,shcol
-    denom(i,nlev) = 1._rtype/ &
-      (1._rtype + cc(i,nlev) + flux(i)*dtime*ggr*rdp_zt(i,nlev))
-    ze(i,nlev) = cc(i,nlev) * denom(i,nlev)
+    d(i,nlev) = 1._rtype - dl(i,nlev) + flux(i)*dtime*ggr*rdp_zt(i,nlev)
   enddo
-
   do k=nlev-1,2,-1
     do i=1,shcol
-      denom(i,k) = 1._rtype/ &
-        (1._rtype + ca(i,k) + cc(i,k) - &
-    ca(i,k) * ze(i,k+1))
-      ze(i,k) = cc(i,k) * denom(i,k)
+      d(i,k) = 1._rtype - du(i,k) - dl(i,k)
```

Comment:
Two things here:
1. You don't need to march backward in k b/c there are no deps on k+/-1 in this loop.
2. You can combine this loop with the next. In C++ that won't be true since the factorization is a separate routine, but in F90 it makes sense to do so.

Reply:
I also included other simplifications in addition to these.